### PR TITLE
[NG] Updates Datagrid DOM Structure

### LINKED ISSUE #1150: If datagrid has horizontal scroll bar, the paging info gets truncated
**Select one ...**  (check one with "x")
```
[x ] bug
[ ] feature request
[ ] enhancement
```

### Actual behavior
Hi All, I'm facing this issue in IE. If datagrid has horizontal scroll bar, the paging info gets truncated.. Note: This is happening only in IE., In Chrome and firefox., it works fine.. Please refer attachment for screenshot..
datagrid_truncation_ie11

### Environment details

* **Angular version:** 2.0.X
4.0.1

* **Clarity version:** 
0.9.7

* **OS and version:** 
windows 10

* **Browser:** [ IE 11 ]

### LINKED ISSUE #1217: When horizontal scrolling is necessary, vertical scroll should be always visible and footer should not use the horizontal scroll
[x] bug
[ ] feature request
[ ] enhancement

### Expected behavior
If a datagrid is too wide to fit in the screen a horizontal scroll bar should appear. This scroll bar should be above the footer/within the table, so that:
1. I can see the vertical scroll all the time (if there is one)
2. I can see the whole footer (it makes no sense to scroll the footer since it has almost no information)

### Actual behavior
Currently the horizontal scroll appears below the footer. Initially I cannot see if there is a vertical scroll, and to use it, I first must scroll to the right, then drag the vertical scroll, then drag the horizontal scroll back to the left, so I can see the first (and presumable most important) column. Also since I can't see neither the vertical scroll, nor the items count (which is on the right side of the footer), I don't even know if there are more items in the datagrid
The same issue would come up if there is paging - the controls for it are on the right side and therefore - hidden.

### LINKED ISSUE #1409: Datagrid loading spinner is not centered
**Select one ...**  (check one with "x")
```
[x ] bug
[ ] feature request
[ ] enhancement
```

### Expected behavior
Datagrid loading spinner should be centered both with and without `clr-dg-action-bar` buttons.

### Actual behavior

The style for `.datagrid-spinner` does not take into account the space added for the action bar. 

### Reproduction of behavior
https://plnkr.co/edit/XAUNqV?p=preview

### Environment details

* **Angular version:** 2.0.X

* **Clarity version:** 
Clarity latest

* **OS and version:** 
n/a

### LINKED ISSUE #1546: Dropdowns in datagrid action bar are misaligned
**Select one ...**  (check one with "x")
```
[x] bug
[ ] feature request
[ ] enhancement
```

### Expected behavior

A dropdown in the datagrid action bar should line up with other buttons.

### Actual behavior

The dropdown is pushed down 1px.

### Reproduction of behavior

* Include a link to the reproduction scenario you created by forking one of the Clarity Plunker Templates: https://embed.plnkr.co/WpEcxC/

### Environment details

* **Angular version:** 4.4

* **Clarity version:** 0.10.6

* **OS and version:** 

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

### LINKED ISSUE #1631: Underlying checkboxex are in front of Dropdown body
```
[x ] bug
[ ] feature request
[ ] enhancement
```

### Expected behavior
Use Dropdown in Datagrid ActionBar.

### Actual behavior
Checkboxes for selecting items are in front of dropdown body

### Reproduction of behavior
https://plnkr.co/edit/3nl7U7yjPXW9cHwI2gTr?p=preview

### Environment details

* **Angular version:** 4.x.x

* **Clarity version:** 0.10.10

* **OS and version:** 

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

### LINKED ISSUE #1675: `clr-dg-row-detail` Components should expand to take the full width of the parent
**Select one ...**  (check one with "x")
```
[x] bug
[ ] feature request
[ ] enhancement
```

### Expected behavior

The `clr-dg-row-detail` component should expand to take the full width of its parent (row) container. 

### Actual behavior
The `clr-dg-row-detail` component is only as wide as its content. 

See the datagrid under Custom Component Row Details in [this](https://plnkr.co/edit/uVsLLzqcgnjTc9SXUyVG?p=preview) plunker for a reproduction scenario. 

### Environment details

* **Angular version:** 4.x.x

* **Clarity version:** `^0.10.12`

* **OS and version:** All

* **Browser:** [all]

# Workaround
For now the user can fix this by adding `flex: 1 1 auto;`to the child thats not expanding to fill its container. 

See this plunker for the workaround: 
https://plnkr.co/edit/uVsLLzqcgnjTc9SXUyVG?p=preview

### PR DESCRIPTION
**Pull request by hippee-lee**
This changes the datagrid templates and updates the DOM structure to support
- better scrolling for content overflow
- future enhancements that need a simpler DOM structure
- Updated css reference images because of the change to top margin when there is no `clr-dg-action-bar` declared

## FIXES
1. Fixes #1409 - Loading spinner not centered
2. Fixes #1217 - Footer should not scroll when overflow-x
3. Fixes #1150 - horizontal scrollbar covers paging info
4. Fixes #1675 - `clr-dg-row-detail` should expand to full width of parent
5. Fixes #1631 - row selection checkboxes are in front of a dropdown in the `clr-dg-action-bar`
6. Fixes #1546 - Dropdowns in actionbar are misaligned

## UNFIXED / KNOWN ISSUES
1. X-axis scroll bars appears when hiding the last column. The fix for this will also fix issue #1248 - It might just be refactoring the resize column handles to be inside the column instead of outside it, it also might need code changes to identify when a column is the last visible and *ngIf the column resize toggle.
2. When datagrid is shorter then a filter popover is tall the content overflows-y and cuts off the bottom og the popover and a y-axis scroll bar appears

## TODO
- [x] Fix failing tests
- [ ] Add Feature Detection util for sticky attribute
- [ ] Add fallback for IE11 that implements a scroll event listener and repositions the column headers
- [x] Confirm with UX the loading indicator position (and action blocking)
- [x] Confirm with UX the top-margin behavior with/without the `clr-dg-action-bar`

Signed-off-by: Matt Hippely <mhippely@vmware.com>

### CODE REVIEW

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 1]
```diff
@@ -3,266 +3,271 @@
...
-        min-width: 100%;
     }
+
     .datagrid {
         display: flex;
-        flex-flow: column nowrap;
+        flex-wrap: wrap;
+        flex: 1 1 auto;
+        width: 100%;
+        margin-top: 0;
```
Comment (adityarb88) [inline src/clr-angular/data/datagrid/_datagrid.clarity.scss]: Why did you add a`margin-top: 0` here? Could this be the reason that we see the diffs in the CSS regression images?

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 2]
```diff
@@ -3,266 +3,271 @@
...
-        min-height: 1px;
-    }
-    .datagrid-body {
-        flex: 1 1 auto;
-        overflow-y: auto;
-        -ms-overflow-style: -ms-autohiding-scrollbar;
 
-        display: flex;
-        flex-direction: column;
+    .datagrid-body .datagrid-row {
```
Comment (adityarb88) [inline src/clr-angular/data/datagrid/_datagrid.clarity.scss]: Do we need to nest the `.datagrid-row` selector under `.datagrid-body`? Is there a case where the `.datagrid-row` is present outside of the `.datagrid-body`? If not, just `.datagrid-row` should be fine here.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 3]
```diff
@@ -3,266 +3,271 @@
...
+        &.datagrid-selected {
+            color: $clr-black;
+            background-color: $clr-global-selection-color;
         }
 
-        &--hidden {
-            display: none;
-        }
-    }
+        .datagrid-action-overflow {
```
Comment (adityarb88) [inline src/clr-angular/data/datagrid/_datagrid.clarity.scss]: Doesn't look like these styles were added in this PR. Is there a reason we rearranged the file? Ideally we should do this kind of cleanup in a separate PR because I can't figure out what was added and what was removed from the diff. Its fine for now but just a note for future PRs.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 4]
```diff
@@ -3,266 +3,271 @@
...
+
         .datagrid-row-detail {
-            display: none;
-        }
-        .datagrid-column, .datagrid-cell {
-            display: table-cell;
-        }
-        .datagrid-column-separator {
-            display: none;
+            &.datagrid-container {
```
Comment (adityarb88) [inline src/clr-angular/data/datagrid/_datagrid.clarity.scss]: Why did we get rid of the `display: none` in the `.datagrid-row-detail` selector?

Comment (hippee-lee) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: It's still there, because the file was rewritten the classes used for computation are at the bottom of the file.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss (hunk 3 above)
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: I had to develop the style in a separate scss file and then copied/pasted it back into the original Basically its been re-written from scratch while porting over the original styles where it made sense.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss (hunk 2 above)
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: This was copied over from the original stylesheet. I moved it out and it seems like the row components are ok. Will move it out from under .datagrid-body.

Will update.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss (hunk 1 above)
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: datagrid inherits from the basic table style that has a margin-top of 1rem. Since the clr-dg-action-bar needs to be 0.5rm above the datagrid I reset it here and added bottom-margin to the action bar. I'm not clear on what the spec is for datagrid top margin and the action-bar.

File: .travis.yml [hunk 5]
```diff
@@ -29,14 +29,13 @@ addons:
       - google-chrome-stable
   artifacts:
     paths:
-      - "$TRAVIS_BUILD_DIR/reports/"
-
+      - $TRAVIS_BUILD_DIR/reports/gemini
```
Comment (gnomeontherun) [inline .travis.yml]: We should also be capturing coverage, so shouldn't need to change this line

File: src/clr-angular/data/datagrid/render/column-resizer.ts [hunk 6]
```diff
@@ -8,17 +8,20 @@ import {AfterViewInit, Directive, ElementRef, EventEmitter, OnDestroy, Output, R
...
 @Directive({selector: "clr-dg-column", providers: [DragDispatcher]})
 export class DatagridColumnResizer implements AfterViewInit, OnDestroy {
     constructor(el: ElementRef, private renderer: Renderer2, private organizer: DatagridRenderOrganizer,
-                private domAdapter: DomAdapter, private dragDispatcher: DragDispatcher) {
+                private domAdapter: DomAdapter, private dragDispatcher: DragDispatcher,
+                private table: TableHeightService) {
         this.columnEl = el.nativeElement;
     }
 
+    // columnHeight: number = 462;
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/render/column-resizer.ts]: Remove

File: src/clr-angular/data/datagrid/datagrid-column.spec.ts [hunk 7]
```diff
@@ -433,7 +435,7 @@ class SimpleDeprecatedTest {
 class SimpleTest {
     comparator: ClrDatagridComparatorInterface<any>|string;
     field: string;
-    sortOrder = ClrDatagridSortOrder.Unsorted;
+    sortOrder = ClrDatagridSortOrder.UNSORTED;
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/datagrid-column.spec.ts]: How did this compile before?!

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 8]
```diff
@@ -340,309 +334,235 @@
...
+            }
 
-                    clr-icon {
-                        vertical-align: middle;
-                        transform: translate3d(0px, -1px, 0);
-                    }
-                }
+            &.datagrid-fixed-column {
+                flex: 0 0 $clr-datagrid-fixed-column-size;
+                // IE10 needs this
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/_datagrid.clarity.scss]: Do we need to support IE10 here?

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 9]
```diff
@@ -340,309 +334,235 @@
...
+                display: none;
             }
-        }
 
-        .datagrid-body .datagrid-row {
-            border-top: $clr-table-borderwidth solid $clr-table-border-color;
-            flex-shrink: 0;
+            &.datagrid-fixed-column {
+                flex: 0 0 $clr-datagrid-fixed-column-size;
+                // IE10 needs this
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/_datagrid.clarity.scss]: Again, IE10?

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 10]
```diff
@@ -3,266 +3,260 @@
...
-            vertical-align: top;
-            background: none;
-            border-bottom: 0;
+        .datagrid-select {
+            min-width: calc(0.375rem - #{$clr-table-borderwidth}); // This fixed an issue that made the cell to wide.
+            .radio label,
+            .checkbox label {
+                display: block;
+                // Override default checkbox styles for width/height
+                min-height: $clr-table-lineheight - $clr-rem-1px;
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/_datagrid.clarity.scss]: nit: space between `$clr-rem - 1px;`

File: scripts/docker-cdc.js [hunk 11]
```diff
@@ -57,7 +57,7 @@ if (program.update) {
 function runGemini(config) {
     shell.exec("webpack");
     let server = shell.exec("node_modules/.bin/lite-server --baseDir=dist" , { async: true });
-    let status = shell.exec("docker run --rm --name=clarity_chrome -d -p 4444:4444 selenium/standalone-chrome@sha256:b899f16b6d963600ef6da8a8dd49e311146033ed66cb5af71eccb78ab378e19a");
+    let status = shell.exec("docker run --rm --name=clarity_chrome -d -p 4444:4444 --shm-size=2g selenium/standalone-chrome@sha256:b899f16b6d963600ef6da8a8dd49e311146033ed66cb5af71eccb78ab378e19a");
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply scripts/docker-cdc.js]: I don't think so, I'm not even sure why that got changed. Will revert.

File: .travis.yml (hunk 5 above)
Comment (hippee-lee) [reply .travis.yml]: Will revert.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss (hunk 8 above)
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: Will remove the comment. We need the min/max widths set.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss (hunk 9 above)
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: Will remove the comment. We need the min/max widths set.

File: scripts/docker-cdc.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (hippee-lee) [reply scripts/docker-cdc.js]: Turns out I am getting alot of these:
`Error response status: 100,  Selenium error: chrome not reachable`
without upping the swap. Going to keep it for now.

File: src/clr-angular/data/datagrid/datagrid-column.spec.ts (hunk 7 above)
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/datagrid-column.spec.ts]: ¯\_(ツ)_/¯

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss (hunk 10 above)
Comment (hippee-lee) [reply src/clr-angular/data/datagrid/_datagrid.clarity.scss]: Didn't see this when I made changes for the review. Perhaps when I squashed it was updated?

**Comment by youdz** [comment]
I can't comment on a screenshot, but the gemini image `gemini/screens/datagrid-compact/datagrid-compact-expandable-rows/columns/chrome.png` isn't compact anymore.

File: src/clr-angular/data/datagrid/animation-hack/row-expand-animation.ts [hunk 12]
```diff
@@ -68,10 +68,9 @@ export class DatagridRowExpandAnimation {
         this.renderOrganizer.scrollbar.next();
         const newHeight = this.domAdapter.computedHeight(this.el.nativeElement);
         this.running = this.el.nativeElement.animate(
-            {height: [this.oldHeight + "px", newHeight + "px"], overflowY: ["hidden", "hidden"], easing: "ease-in-out"},
-            {duration: 200});
+            {height: [this.oldHeight + "px", newHeight + "px"], easing: "ease-in-out"}, {duration: 200});
```
[the remark this answers is not in the document]
Comment (youdz) [reply src/clr-angular/data/datagrid/animation-hack/row-expand-animation.ts]: We're animating from a dynamic height to a dynamic height*, so there is no way to do it in CSS. At the time, Angular animations didn't support it either, so I had to do it manually.

* The old height is the height of the collapsed row, which is determined solely by its content (wrapping text, ...). The new height is the height of the row + its detail, once again depending on its content.

File: src/clr-angular/data/datagrid/datagrid-column.spec.ts (hunk 7 above)
Comment (youdz) [reply src/clr-angular/data/datagrid/datagrid-column.spec.ts]: 😂

File: src/clr-angular/data/datagrid/datagrid-hideable-column.spec.ts [hunk 13]
```diff
@@ -13,11 +13,13 @@ import {FiltersProvider} from "./providers/filters";
 import {Page} from "./providers/page";
 import {Sort} from "./providers/sort";
 import {StateDebouncer} from "./providers/state-debouncer.provider";
+import {TableHeightService} from "./providers/table-height.service";
 import {DomAdapter} from "./render/dom-adapter";
 import {DatagridRenderOrganizer} from "./render/render-organizer";
 
-const PROVIDERS_NEEDED =
-    [Sort, FiltersProvider, DatagridRenderOrganizer, DomAdapter, DragDispatcher, Page, StateDebouncer];
+const PROVIDERS_NEEDED = [
```
Comment (youdz) [inline src/clr-angular/data/datagrid/datagrid-hideable-column.spec.ts]: We really need to set this on the test context at the top level of the Datagrid specs. This is getting silly. 😂

File: src/clr-angular/data/datagrid/animation-hack/row-expand-animation.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (Shijir) [reply src/clr-angular/data/datagrid/animation-hack/row-expand-animation.ts]: Accessing `nativeElement.animate` also means touching the dom, right?  What if we just apply `transition: height 0.2s ease-in-out` to the element that's animating, and then use Renderer to create that toggle animation effect by setting the height to `newHeight + "px"` and `this.oldHeight + "px"`.

[the remark this answers is not in the document]
Comment (youdz) [reply src/clr-angular/data/datagrid/animation-hack/row-expand-animation.ts]: Feel free to try it on a different PR, but it doesn't have anything to do with this one. And I'm pretty skeptical about using CSS on that, my first try would be to use the latest versions of `@angular/animations` which provide `* => *` height animations.

File: src/clr-angular/data/datagrid/providers/table-height.service.ts [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+ * @description
+ * Internal datagrid service that holds a reference to the clr-dg-table element.
+ */
+@Injectable()
+export class TableHeightService {
+    public tableRef: ElementRef;
+
+    set tableElement(table: ElementRef) {
+        this.tableRef = table;
+    }
```
Comment (youdz) [inline src/clr-angular/data/datagrid/providers/table-height.service.ts]: Could you add a `getHeight()` method or `get height()` setter to this class? It seems like the only way you're using this everywhere is by chaining to the native elements and computed height, so you should just expose directly a numeric height here.

On top of that, you have a setter on a public property, which is pretty useless in Typescript/Javascript. 😛 
You should just remove the setter.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 15]
```diff
@@ -337,312 +371,212 @@
...
         background-color: $clr-thead-bgcolor;
-        border-top: (2 * $clr-default-borderwidth) solid $clr-table-footer-border-top-color;
+        border-top: $clr-default-borderwidth solid $clr-table-footer-border-top-color;
         border-radius: 0 0 $clr-global-borderradius $clr-global-borderradius;
+        border-left: $clr-table-borderstyle;
+        border-bottom: $clr-table-borderstyle;
+        border-right: $clr-table-borderstyle;
+
+        .datagrid-footer-left, .datagrid-footer-right {
+            flex: 1;
```
Comment (youdz) [inline src/clr-angular/data/datagrid/_datagrid.clarity.scss]: See my previous PR on the footer, I'm positive you want the `.datagrid-footer-left` to _not_ be flexible (`flex: 0 0 auto)`) and the `footer-right` to be the one filling up the space.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 16]
```diff
@@ -3,266 +3,300 @@
...
         display: flex;
         flex-flow: column nowrap;
-    }
-
-    .datagrid-overlay-wrapper {
-        display: flex;
-        flex-direction: row;
-        flex: 0 auto; // Same as flex: initial. It sizes the item based on width/height properties or its content.
+        height: auto;
+        position: relative; // for the .datagrid-spinner to work.
```
Comment (youdz) [inline src/clr-angular/data/datagrid/_datagrid.clarity.scss]: Damnit, I didn't see this `position: relative`.  That will cut off any filters popover, action overflow menus and column picker... I'm not sure I want to block the PR on this right now, but can you put a `TODO` or `FIXME` on this so we can re-investigate at the "end" of this whole topic branch?

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 17]
```diff
@@ -890,12 +825,63 @@
...
     }
 
-    @include fixForMsEdge {
-        .datagrid .datagrid-column .datagrid-filter-toggle {
-            // MSEdge doesn't round the rem units properly
-            width: 0.59rem;
-            height: 0.59rem;
-            margin-top: -1px;
+    // Used in table-renderer.ts in noLayout with NO_LAYOUT_CLASS.
+    .datagrid-no-layout {
```
Comment (youdz) [inline src/clr-angular/data/datagrid/_datagrid.clarity.scss]: If you don't mind, can you put the original comment too, to explain this is just for performance optimization when rendering? Here is is:
```typescript
// Class only applied temporarily when preforming internal operations,
// to improve rendering performance.
```

Feel free to edit it, obviously, all that matters is that the comment gives the "why" of these styles.

**Comment by hippee-lee** [comment]
That makes two of us `spinning` (instead of pinning)? 
I'll be pushing an update to address your feedback. Am not planning to answer each individually though.

File: src/clr-angular/data/datagrid/_datagrid.clarity.scss [hunk 18]
```diff
@@ -714,10 +653,29 @@
             }
         }
     }
-
-    .datagrid-foot-description {
+    .datagrid-footer-select.checkbox {
         display: block;
-        white-space: nowrap;
+        line-height: inherit;
+        flex: 1 1 (1.5rem - (2 * $clr-rem-1px)); // can't use calc() in flex. #thanksIE!
```
Comment (youdz) [inline src/clr-angular/data/datagrid/_datagrid.clarity.scss]: I think this should be `flex: 0 0 ...`?
